### PR TITLE
Run WSL from Removable Media

### PR DESCRIPTION
**Pull request by magliocchetti**
## Summary

This PR implements support for portable WSL distributions that can run directly from removable media (USB drives, external disks, etc.) without requiring local disk installation. This addresses feature request #12804 and enables true portability for WSL environments.

## Problem Statement

Currently, WSL forces all distributions to be installed on local disks, creating two major limitations:

1. **Lack of true portability** - Developers cannot maintain self-contained WSL environments on removable drives
2. **Host system contamination** - Temporary usage on foreign computers leaves persistent installations and user data traces

## Solution Overview

This PR introduces a complete portable WSL workflow with three new commands and supporting infrastructure:

#### 1. `wsl --mount-removable <path> [options]`

Mount a portable WSL distribution from removable media.

**Options:**
- `--distro <name>` or `-d <name>`: Override distribution name
- `--temporary` or `-t`: Mark as temporary (registry-tracked for future auto-cleanup)
- `--allow-fixed` or `-f`: Allow mounting from fixed drives (for development/testing)

**Examples:**
```bash
# Mount from USB drive (default behavior)
wsl --mount-removable E:\wsl\ --distro PortableUbuntu

# Mount from fixed drive for testing (requires explicit flag)
wsl --mount-removable C:\dev\wsl\ --distro TestPortable --allow-fixed
```

#### 2. `wsl --import <name> <location> <file> --portable [--allow-fixed]`

Create a new portable distribution on external storage.

**Examples:**
```bash
# Create on USB drive (default behavior)
wsl --import UbuntuPortable F:\wsl\ubuntu ubuntu-22.04.tar.gz --portable --vhd

# Create on fixed drive for testing (requires explicit flag)
wsl --import TestPortable C:\dev\wsl\test ubuntu-22.04.tar.gz --portable --vhd --allow-fixed
```

#### 3. `wsl --unmount-removable <distro-name>`

Cleanly unmount a portable distribution and remove system traces.

**Example:**
```bash
wsl --unmount-removable PortableUbuntu
```

### Architecture

```
[USB Drive]
├── wsl-portable.json     # Distribution metadata
└── distro.vhdx           # Distribution filesystem
```

### Metadata Format

The `wsl-portable.json` file contains distribution configuration:

```json
{
    "Name": "PortableUbuntu",
    "FriendlyName": "Ubuntu 22.04 Portable",
    "VhdxPath": "ubuntu.vhdx",
    "Version": 2,
    "DefaultUid": 1000,
    "IsPortable": true
}
```

### Key Components

1. **PortableDistribution.h/cpp**
   - Core portable WSL functionality
   - Metadata read/write operations
   - **Security-focused path validation** (removable media by default)
   - Distribution mounting/unmounting logic
   - `allowFixed` parameter for development/testing flexibility

2. **Enhanced WslClient.cpp**
   - Command routing for new options
   - Integration with existing WSL infrastructure
   - **Proper localization** using `Localization::MessagePortable*()` calls
   - Portable distribution tracking
   - `--allow-fixed` flag support

3. **Localization (Resources.resw)**
   - 7 new localized message strings
   - Replaced temporary inline messages with proper localization

4. **Registry Integration**
   - Portable distributions marked in registry
   - Base path stored for cleanup
   - **Temporary flag tracking** (`IsTemporary` DWORD)
   - Transient registration support

### Path Handling

- Supports standard paths: `E:\wsl\`
- Supports Volume GUID paths: `\\?\Volume{UUID}\wsl\`
- **Automatic drive type detection** using Windows `GetDriveTypeW()` API
- **Security-focused validation**: Restricts to `DRIVE_REMOVABLE` by default
- **Explicit opt-in for fixed drives**: `--allow-fixed` flag required
- Helpful error messages guide users to correct flag usage

## Benefits

✅ **True Portability**: Carry complete Linux environments on USB drives  
✅ **Zero Local Footprint**: No contamination of host system  
✅ **Clean Removal**: Automatic cleanup on unmount  
✅ **Easy Sharing**: Share pre-configured environments  
✅ **Secure by Default**: Removable media only; fixed drives require explicit flag  
✅ **Developer Flexibility**: `--allow-fixed` enables testing without USB drives  
✅ **Temporary Mounts**: Registry-tracked for future auto-cleanup features  

### Creating a Portable Distribution

```bash
# Method 1: Import from tar file as portable (USB drive)
wsl --import MyPortable E:\wsl\ ubuntu-22.04.tar.gz --portable --vhd

# Method 2: Export existing, then import as portable
wsl --export Ubuntu ubuntu-backup.tar.gz
wsl --import PortableUbuntu F:\wsl\ ubuntu-backup.tar.gz --portable --vhd

# Method 3: Create on fixed drive for testing (explicit opt-in)
wsl --import TestPortable C:\dev\wsl\ ubuntu-22.04.tar.gz --portable --vhd --allow-fixed
```

### Using a Portable Distribution

```bash
# Mount the portable distribution
wsl --mount-removable E:\wsl\ --distro MyPortable

# Use it like any other distribution
wsl -d MyPortable

# When done, clean unmount
wsl --unmount-removable MyPortable
```

### Sharing Portable Distributions

Simply copy the entire portable directory (containing `wsl-portable.json` and `.vhdx` file) to another USB drive or share with others.

### Files Modified

- `src/windows/inc/wsl.h` - Added command constants including `--allow-fixed` flags
- `src/windows/common/WslClient.cpp` - Command handlers, routing, localization integration
- `src/windows/common/WslClient.h` - Function declarations
- `src/windows/common/CMakeLists.txt` - Added new source files
- `localization/strings/en-US/Resources.resw` - Added 7 new localized message strings

### Files Added

- `src/windows/common/PortableDistribution.h` - Portable WSL API
- `src/windows/common/PortableDistribution.cpp` - Implementation

### Backward Compatibility

- All existing WSL commands and functionality remain unchanged
- New commands are opt-in via explicit flags
- No impact on non-portable distributions
- Existing registry structure preserved

## Testing Recommendations

1. **Basic Functionality**
   - Create portable distribution on USB drive
   - Mount and use portable distribution
   - Unmount and verify cleanup

2. **Security & Path Validation**
   - Verify removable media works without flags
   - Confirm fixed drives are **blocked by default**
   - Test `--allow-fixed` flag enables fixed drives
   - Test standard paths (E:\wsl\)
   - Test Volume GUID paths (\\?\Volume{...}\)
   - Verify error messages guide users to `--allow-fixed`

3. **Localization Testing**
   - Verify all 7 new messages display correctly
   - Confirm error messages are properly localized

4. **Temporary Flag Tracking**
   - Mount with `--temporary` flag
   - Verify `IsTemporary=1` in registry
   - Confirm cleanup removes temporary flag

5. **Multi-Device Testing**
   - Create on one machine, use on another
   - Test drive letter changes
   - Verify portability across Windows versions

6. **Edge Cases**
   - Mount already-mounted portable distribution
   - Unmount non-portable distribution
   - Handle missing metadata file
   - Handle corrupted VHDX
   - Test fixed drive without `--allow-fixed` (should fail with helpful message)

7. **Registry Cleanup**
   - Verify portable flag in registry
   - Verify temporary flag when used
   - Confirm cleanup after unmount
   - Test multiple portable distributions

## Future Enhancements (Out of Scope)

- **Automatic cleanup on drive ejection** (foundation laid with temporary flag tracking)
- **Auto-cleanup on reboot/logout** for temporary distributions
- Encryption support for portable distributions (BitLocker integration)
- Compression options for smaller footprint
- GUI integration in WSL Settings app
- Auto-mount on drive insertion
- Service-side temporary flag monitoring

## Documentation Updates Needed

- User documentation for new commands
- Examples in WSL documentation
- Troubleshooting guide for portable distributions
- Document `--allow-fixed` flag behavior and security rationale

### Creating a Portable Distribution

```
PS C:\> wsl --import PortableUbuntu E:\wsl\ ubuntu.tar.gz --portable --vhd
Installing: 100.0%
Portable distribution 'PortableUbuntu' created successfully at 'E:\wsl\'.
The operation completed successfully.
```

### Mounting a Portable Distribution

```
PS C:\> wsl --mount-removable E:\wsl\ --distro PortableUbuntu
Portable distribution 'PortableUbuntu' successfully mounted from 'E:\wsl\ext4.vhdx'.
The operation completed successfully.

PS C:\> wsl -d PortableUbuntu
user@DESKTOP-ABC:~$ df -h
Filesystem      Size  Used Avail Use% Mounted on
/dev/sdc       1007G  123G  833G  13% /
```

### Unmounting a Portable Distribution

```
PS C:\> wsl --unmount-removable PortableUbuntu
Portable distribution 'PortableUbuntu' successfully unmounted.
The operation completed successfully.
```

## Checklist

- [x] Code follows WSL coding standards
- [x] New functionality is properly abstracted
- [x] Error handling implemented
- [x] Backward compatibility maintained
- [x] No breaking changes to existing functionality
- [x] Commit message follows convention
- [ ] Documentation updated
- [ ] Tests added

### CODE REVIEW

File: src/windows/common/PortableDistribution.cpp [hunk 1]
```diff
@@ -0,0 +1,435 @@
...
+        distroName,
+        version,
+        sourceFileHandle.get(),
+        portablePath.c_str(),
+        flags | LXSS_IMPORT_DISTRO_FLAGS_VHD);
+
+    // Unregister immediately as we only wanted to create the VHDX
+    try
+    {
+        service.UnregisterDistribution(&guid);
```
Comment (OneBlue) [inline src/windows/common/PortableDistribution.cpp]: Doesn't unregistering the distribution delete the VHDX as well on this code path ?

**Comment by OneBlue** [comment]
Closing since we haven't heard back from that author. Feel free to reopen in the future.
